### PR TITLE
Make Interpreter.Object and friends into (non-static) inner classes of Interpreter

### PR DESCRIPTION
**Pull request by cpcallen**
The type error fixes in 3d513c7 are probably just temporary until I can figure out how to use @interface / @implements correctly.  The way I've done it here is not very nice (but does fix the warnings!)

The main thing that requires consideration is the refactor in 26585ca.

### CODE REVIEW

File: server/interpreter.js [hunk 1]
```diff
@@ -2246,6 +1979,393 @@ Interpreter.prototype.pushNode_ = function(node) {
 };
 
 ///////////////////////////////////////////////////////////////////////////////
+// Types representing JS objects
+///////////////////////////////////////////////////////////////////////////////
+
+// This is a bunch of boilerplate that serves two purposes:
+// 
+// * First, by declaring these types as if they were on
+//   Interpreter.prototype we can get the closure compiler to type
```
Comment (NeilFraser) [inline server/interpreter.js]: Closure Compiler

File: server/interpreter.js [hunk 2]
```diff
@@ -2246,6 +1979,393 @@ Interpreter.prototype.pushNode_ = function(node) {
...
+//   check use of them for us.
+//
+// * Second, for whatever reason these declarations seem to create a
+//   small performance improvement.
```
Comment (NeilFraser) [inline server/interpreter.js]: ... in V8.

File: server/interpreter.js (hunk 1 above)
Comment (cpcallen) [inline server/interpreter.js]: Done.

File: server/interpreter.js (hunk 2 above)
Comment (cpcallen) [inline server/interpreter.js]: Done.

**Pull request merged**